blog/_posts/2015-12-2-includeos-is-now-open-source.md

```diff
@@ -0,0 +1,38 @@
+
+## We want to make Node++
+IncludeOS is pretty similar to Node.js in a few respects: It's single threaded at the moment, but highly efficient due to non-blocking I/O. This means that you'll have to use a callback-based programming style, just like with Javascript. The events you'll be subscribing to are all rooted in a physical interrupt - in fact, you can add callbacks to interrupts directly, if you want.  An interesting thing to keep in mind is that I/O in modern hardware is non-blocking by default, with Direct Memory Access (DMA). A network packet arriving into IncludeOS will result in an interrupt, but only after the whole packet is written directly into memory by the device. At that point, IncludeOS will defer the interrupt, and as soon as the CPU is done with whatever it was doing, it will call the callback delegate (possibly a lambda) of whoever's subscribing to that interrupt. That delegate can again use the same technique, to fire events higher up the stack. In our view, this kind of event-based programming fits hardware better than sequential, blocking programming. Also, GUI-programming has always been like this, so the techniques are well known. 
+
+### We're making a github-based package manager
+One very important feature we'll be needing in order to make IncludeOS grow, is a package manager. We're making one much like [npm](https://www.npmjs.com/), where anyone can create a package as a github repository, and then register that package in a central location for others to use. The idea is to have `$ include http` on the terminal, result in the possibility of writing `#include <http>` inside your IncludeOS service. This will obviously take some time, but hopefully we have something working by summer 2016.
+
+## Networking from scratch
+One thing you'll notice is that our network stack is currently being written from scratch. Are we crazy? Possibly. We obviously considered - and wanted to - port the stack from somewhere else, but not being intimately familiar with any existing kernel, I found that to be even harder than just doing it from scratch. I got a lot of help from reading the nicely annotated and cross-linked [SanOS source code](http://www.jbox.dk/sanos/source/), and once the virtio network driver was working, the rest wasn't nearly as hard as one might think - at least up to now. Clearly it will take time to make it complete, but the efforts it took to get form a driver, up to rudimentary TCP was really insignificant compared to the challenges we've faced when trying to integrate third-party libraries such as newlib (Standard C) and libc++ (Standard C++). And - getting to read RFC's and implementing the TCP handshake from scratch has really been a blast - a highly recommended exercise.
```

Comment:
Knowing other bits of the unikernel ecosystem, I immediately wondered if Rump Kernels could have been useful here.

Also, nice to read that someone else also saw the value in starting from scratch with RFCs for core protocols.  I'm sure there are plenty of stories to exchange with the HaLVM and MirageOS crews :)

minor typos:
- form a driver -> from a driver
- And - getting to read RFC's -> Getting to read RFCs